astroquery/esasky/tests/test_esasky_remote.py

```diff
@@ -1,352 +1,281 @@
 
-    def test_esasky_get_maps(self):
-        download_directory = "ESASkyRemoteTest"
-        if not os.path.exists(download_directory):
-            os.makedirs(download_directory)
-
-        file_path = os.path.join(download_directory, 'ISO-IR')
+    def test_esasky_get_maps(self, tmp_path):
+        file_path = tmp_path / 'ISO-IR'
 
-        all_maps = ESASkyClass.query_object_maps(position="M51")
-        iso_maps = ESASkyClass.query_object_maps(position="M51", missions='ISO-IR')
+        all_maps = ESASky.query_object_maps(position="M51")
+        iso_maps = ESASky.query_object_maps(position="M51", missions='ISO-IR')
         # Remove a few maps, so the other list will have downloadable ones, too
         iso_maps['ISO-IR'].remove_rows([0, 1])
-        ESASkyClass.get_maps(iso_maps, download_dir=download_directory)
+        ESASky.get_maps(iso_maps, download_dir=tmp_path)
         assert len(os.listdir(file_path)) == len(all_maps['ISO-IR']) - 2
 
         iso_maps2 = dict({'ISO-IR': all_maps['ISO-IR'][:2]})
-        ESASkyClass.get_maps(iso_maps2, download_dir=download_directory)
+        ESASky.get_maps(iso_maps2, download_dir=tmp_path)
         assert len(os.listdir(file_path)) == len(all_maps['ISO-IR'])
 
-        shutil.rmtree(download_directory)
-
     def test_esasky_query_region_spectra(self):
-        result = ESASkyClass.query_region_spectra(position="M51", radius="5 arcmin")
+        result = ESASky.query_region_spectra(position="M51", radius="5 arcmin")
         assert isinstance(result, TableList)
 
     def test_esasky_query_object_spectra(self):
-        result = ESASkyClass.query_object_spectra(position="M51")
+        result = ESASky.query_object_spectra(position="M51")
         assert isinstance(result, TableList)
 
     @pytest.mark.bigdata
-    def test_esasky_get_spectra(self):
-        download_directory = "ESASkyRemoteTest"
-        if not os.path.exists(download_directory):
-            os.makedirs(download_directory)
-
-        missions = ESASkyClass.list_spectra()
-        # HST-IR has no data, LAMOST does not support download
-        missions = [mission for mission in missions if mission not in ("HST-IR", "LAMOST")]
-        ESASkyClass.get_spectra(position="M1", missions=missions, download_dir=download_directory)
-
-        for mission in missions:
-            file_path = os.path.join(download_directory, mission)
-            assert file_path[:16] == "ESASkyRemoteTest"
-
-        shutil.rmtree(download_directory)
-
-    def test_esasky_get_spectra_small(self):
-        download_directory = "ESASkyRemoteTest"
-        if not os.path.exists(download_directory):
-            os.makedirs(download_directory)
-
+    @pytest.mark.parametrize("mission", ["XMM-NEWTON", "Chandra", "IUE", "HST-UV",
+                                         "HST-OPTICAL", "ISO-IR", "Herschel"])
+    def test_esasky_get_spectra(self, tmp_path, mission):
+        # Not all missions are tested here:
```

Comment:
This can change. Could we maybe keep the logic of generating the missions list with a query outside this test, then can use that list to populate the parametrization?  Or any other way that would be somewhat sensitive for changing supported missions.

Reply:
I had the same thought and put the

```python
    missions = ESASky.list_spectra()
```

call outside the test, to generate the list, and then parametrize based on the list.  The problem is that this `ESASky.list_spectra()` runs on test collection, which means it queries remote data just to collect tests names, which causes an error, when `--remote-data=any` is not set, i.e. in the normal CI of this repo.  Setting module-level `pytestmark = pytest.mark.remote_data` doesn't change this.  The only way to avoid it would be to have some other type of exclude for the test*_remote.py files when `--remote-data=any` is not specified.

Anway, I couldn't figure out a way to get around it, unfortunately.

On the flip side, my updated version of the test is robust to new missions being added with out any observations of M1.  😅

Reply:
Have you tried using a [parametrized fixture](https://docs.pytest.org/en/stable/how-to/fixtures.html#parametrizing-fixtures)?

Reply:
Hey that works!  🎉

Reply:
Correction, it does not.  It needs the list to know how many tests are collected so it can report which ones were skipped.  It can't do this if it needs remote data:

```
E   OSError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: 193.147.153.113
```

Reply:
It looks like you might have to add an if-statement in the fixture that checks the remote-data setting. Something like this:
```python
@pytest.fixture
def foo(request):
    if request.config.getoption("remote-data") != "none":
        return actual_remote_data
    else:
       return local_data_if_possible_else_None_or_something
```

Reply:
My last comment isn't actually helpful in this situation, what you really need is [using marks with parametrized fixtures](https://docs.pytest.org/en/stable/how-to/fixtures.html#using-marks-with-parametrized-fixtures).

Reply:
What about having a list defined then at the top and both used in the parametrisation here (as it's now), and in a separate test to assert it doesn't have any extra or missing missions?

Reply:
But then it also points beyond the scope of this given pr.

Reply:
The idea by @bsipocz should indeed work, but if some set of parameters is used in more than one test then it would be good to set up a parametrized fixture so that the same parameter list would not have to be written multiple times.

Reply:
Thanks for the suggestions.  I like your idea @bsipocz.  That will work well.

Unfortunately, all the various parametrized missions for these tests are very specific to the tests themselves, so don't scale out to multiple tests.  I.e. a particular Messier galaxy has imaging or spec data in most of the missions but not all.  So the list of missions with data will be different from one test to the next.

Reply:
I guess what I'm saying is that for test stability and to work with parametrization, and for clarity in the tests, it would be good to have the tests have a fixed list of missions that are known to have data of the requested type for the requested object in the test.  Each test is different, and this list of missions varies from test to test.

Separately it would be good to have a test that gives the current state of missions in a fixed list and then checks to make sure that it is the same as what `ESASky.list_spectra()` and `ESASky.list_maps()` return.  Those tests would fail when a new mission is added, and it may be good to update the fixed lists in the tests.  But I guess that is beyond the scope of this PR and maybe requires some expertise from ESA.

Reply:
Btw, there's a test close to this here

https://github.com/astropy/astroquery/blob/404fa410820bc97a27fcfa18d1f10e9400651374/astroquery/esasky/tests/test_esasky_remote.py#L326-L327

So maybe do a `set()` comparison between the fixed and queried lists?  And do this for all 4 `list_foo()` methods?

Reply:
sets are great indeed for this. If you could open a separate issue for these we could directly ping Henrik about them.

Reply:
@jdavies-st Many thanks for tidying up the ESASky module! I'll add an item to the ESASky backlog to look into making these tests more robust and at the same time finding any potential issues with newly added missions. Speaking of, I should fix the JWST issue right away.